Add user verification for admin commands

galgeek opened the pull request:
hi @whimboo !

here's a patch for issue #28. I've run some quick tests on it, seems ok.
this could perhaps still be tidied up some, but it is far better than the code it replaces in the first patch.
I've added some error-checking and tweaked the adminhelp array to support all this.

whimboo commented on bot.js:
```diff
@@ -126,73 +126,73 @@ client.addListener('message', function(from, to, message){
+            case ":addAdmin":
+              if (command[1]){
+                testDayAdmins.push(command[1]);
+                client.say(from, 'Test Day admins are now ' + testDayAdmins.toString());
```
Btw. have you tried to use the 0.3.5 version of the plugin to verify that this problem is still persistent? Maybe even check the latest code. Otherwise the replacement sounds good. We might indeed want to base on a node.js library which already implements basic bot logic. We may wanna have a look which libraries exist and what their usage number is.

whimboo commented on bot.js:
```diff
@@ -126,72 +127,95 @@ client.addListener('message', function(from, to, message){
-  if (message.search(':adminhelp') === 0){
+  var command = message.split(" ");
+  if (command[0] in adminhelp){ // admin command
     if (testDayAdmins.indexOf(from) >= 0){
```
Lets negate this check and return early so that we can reduce the nested if/else conditions.

whimboo commented on bot.js:
```diff
@@ -126,72 +127,95 @@ client.addListener('message', function(from, to, message){
+                      client.say(from, 'Test Day admins are now ' + testDayAdmins.join(", "));
+                    } else {
+                      client.say(from, "Sorry! " + command[1] + " is not logged in with a registered nick.");
+                    }
```
I would love if you could restructure this block to return early if conditions are not met. Those if/else conditions are nested way too much.

whimboo replied:
Also personally I do not think that a person needs to be in the channel at the time he/she gets added.

whimboo commented on bot.js:
```diff
@@ -10,6 +10,7 @@ var ircServer = config.server,
       autoRejoin: config.autoRejoin,
     },
     client = new irc.Client(ircServer, nick, options),
+    botChannel = config.channels[0];
```
Where does this change come from? Please remove given that it has nothing to do with this issue.

galgeek commented:
hi @whimboo 

I've restructured this listener to decrease nesting. 

I've also backed out the checks on arguments to :addAdmin and :addHelper. Since we already check for logged on user of a registered nick and membership in the admin list before we run any admin command, it doesn't add much security to check these arguments, too. Moreover, it's not at all clear how we might verify a nick that's not currently signed in; that should probably be a separate issue.

thank you for your feedback!

galgeek commented:
hi @whimboo 

I see some possibilities, in /help ... output from the Mozilla IRC server, for doing additional validation of arguments to :addAdmin/:addHelper, in particular, the /check command available to operators (though apparently not to *all* channel operators, so I can't test them with op on the #testdaybotTest channel), and there may be others. Maybe you know more?

I do think this should be a separate issue.

whimboo commented:
> Moreover, it's not at all clear how we might verify a nick that's not currently signed in; that should probably be a separate issue.

Right now I think we should not allow the action and abort it. Only identified users should be able to get added.

whimboo commented:
Also not exactly sure what you mean with the /help output. Maybe you can clarify?

galgeek commented:
Hi @whimboo 

Mozilla's IRC server has an operator command, `/check <nick>` that looks like it might provide info on nicks that are not currently signed in, or not using our channel. If we want to allow :addAdmin to add a user with a registered nick who is not currently signed in, we need an alternative to node-irc's `Client.whois` to check, since that works only for signed in users.

galgeek commented:
Hi @whimboo 
I’ve rebased on updated master, and removed one more level of nesting.
*I used just `git rebase master` but still had to push +branch.*
Thank you again for all your feedback and help!

whimboo commented:
Hm, `/check` is really for operators of the IRC network. So no-one of us will have access to it.

whimboo commented on bot.js:
```diff
@@ -147,74 +147,82 @@ client.addListener('message', function(from, to, message){
+
+  client.whois(from, function(whoisinfo){
+    if (!(whoisinfo.accountinfo && whoisinfo.accountinfo.search('is logged in as') >= 0)){
+      client.say(from, "Sorry! You're not logged in with a registered nick.");
```
You can also return early here.

nit: something I have seen a lot and which might be helpful to you... opening `{` brackets will also get a blank before. You might want to have a look at our general Mozilla coding styles: https://developer.mozilla.org/en-US/docs/Mozilla/Developer_guide/Coding_Style

whimboo commented on bot.js:
```diff
@@ -147,74 +147,82 @@ client.addListener('message', function(from, to, message){
+              client.say(from, "Next Test Day's topic is " + topic);
+            } else {
+              client.say(from, "Need some help? " + adminhelp[command[0]]);
+            }
```
I see that lot of your code has overlap in the same areas. I would propose that you really finish those PRs which have code changes in different areas in that file. That will save you and me a lot of rebase and update time.

whimboo commented:
This looks fine to me. I will get it merged in a minute. Thanks Barbara!
